### PR TITLE
`tail`: Refactor handling of warnings and early exits

### PR DESCRIPTION
**Pull request by Joining7943**
Refactor

* printing of warnings because of problematic configuration by the user on the command line
* handling of early exits. 

This pr also adds the missing warning because of `--follow` being ineffective under some special circumstances.

The warnings and early exits are organized in dedidacted methods within the `Settings` struct providing:
* Clearer, less complex control flow and overall code structure
* `args::Settings::from` has no side effects anymore and the methods sole purpose is parsing and storing the initial configuration provided by the user
* It's simpler now to produce the warnings in the same order as gnu's tail does
* Early exits are organized in a more prominent place in the main module `tail`
* Fixes an issue with a warning which was printed multiple times instead of only once at the start of `tail`

This pr also adds more tests which originate from the main refactoring pr #3952. A lot of these tests (besides the ones testing the `warning` functionality) show some broken functionality of `tail` or a bug and are therefore (partly) `disabled_until_fixed`, here. However, many of them were fixed on the way of #3952 and are enabled there.

There's also some additional code cleanup:
* Remove `InputService` and move some of its functionality into `Settings`
* Remove unneeded `stdin_is_pipe_of_fifo` function

### CODE REVIEW

File: src/uu/tail/src/args.rs [hunk 1]
```diff
@@ -242,6 +237,82 @@ impl Settings {
...
+        }
+
+        if let Some(check) = no_pid_support {
+            check
+        } else {
+            CheckResult::Ok
+        }
+    }
+
+    pub fn verify(&self) -> VerificationResult {
```
Comment (tertsdiepraam) [inline src/uu/tail/src/args.rs:295]: I like that you refactored `verify` into a separate function. However, `VerificationResult` feels like a `UResult<()>` with extra steps though, with the disadvantage that you can use `?` on it. The mapping is basically:
- `VerificationResult::Ok` -> `Ok(())`
- `VerificationResult::Exit` -> [`ExitCode`](https://docs.rs/uucore/latest/uucore/error/struct.ExitCode.html).
- `VerificationResult::Fatal` -> `USimpleError`

File: src/uu/tail/src/args.rs [hunk 2]
```diff
@@ -242,6 +237,82 @@ impl Settings {
...
+        }
+
+        let mut no_pid_support: Option<CheckResult> = None;
+        if self.pid != 0 {
+            if self.follow.is_none() {
+                show_warning!("PID ignored; --pid=PID is useful only when following");
+            } else if !platform::supports_pid_checks(self.pid) {
+                show_warning!("--pid=PID is not supported on this system");
+                no_pid_support = Some(CheckResult::NoPidSupport);
+            }
```
Comment (tertsdiepraam) [inline src/uu/tail/src/args.rs:269]: I think this warning and the corresponding `settings.pid = 0` should stay in the argument parsing, because then there is a single point where `pid` is set, which I find more important than grouping all the warnings. Simplifies this function because it's sole purpose is now just printing warnings and it doesn't have to return anything.

File: src/uu/tail/src/follow/watch.rs [hunk 3]
```diff
@@ -91,17 +90,21 @@ pub struct WatcherService {
...
+
     pub watcher_rx: Option<WatcherRx>,
     pub orphans: Vec<PathBuf>,
     pub files: FileHandling,
+
+    pub pid: platform::Pid,
 }
 
-impl WatcherService {
+impl Observer {
```
Comment (tertsdiepraam) [inline src/uu/tail/src/follow/watch.rs:101]: Any particular reason for this name change?

File: src/uu/tail/src/platform/unix.rs [hunk 4]
```diff
@@ -45,7 +45,9 @@ pub fn supports_pid_checks(pid: self::Pid) -> bool {
 fn get_errno() -> i32 {
     Error::last_os_error().raw_os_error().unwrap()
 }
+
 #[inline]
+#[allow(dead_code)]
```
Comment (tertsdiepraam) [inline src/uu/tail/src/platform/unix.rs:50]: Please don't do this. These annotations are hard to remove again because they don't report when they are not necessary anymore.

File: src/uu/tail/src/paths.rs [hunk 5]
```diff
@@ -5,24 +5,14 @@
...
-
 use uucore::error::UResult;
 
-use crate::args::Settings;
-use crate::text;
-
-//  * This file is part of the uutils coreutils package.
-//  *
-//  * For the full copyright and license information, please view the LICENSE
-//  * file that was distributed with this source code.
```
Comment (tertsdiepraam) [inline src/uu/tail/src/paths.rs:24]: Nice :)

File: src/uu/tail/src/follow/watch.rs [hunk 6]
```diff
@@ -477,7 +480,7 @@ pub fn follow(mut watcher_service: WatcherService, settings: &Settings) -> UResu
 
         // If `--pid=p`, tail checks whether process p
         // is alive at least every `--sleep-interval=N` seconds
-        if settings.follow.is_some() && settings.pid != 0 && process.is_dead() {
+        if settings.follow.is_some() && observer.pid != 0 && process.is_dead() {
```
Comment (tertsdiepraam) [inline src/uu/tail/src/follow/watch.rs:483]: Why this change from `settings.pid` to `observer.pid`?

File: src/uu/tail/src/args.rs [hunk 7]
```diff
@@ -242,6 +237,82 @@ impl Settings {
...
+        }
+
+        if self.follow.is_some() && !self.default_stdin && self.has_stdin() {
```
Comment (tertsdiepraam) [inline src/uu/tail/src/args.rs:272]: What's the test case for this warning? I can't think of any reason why it should only apply with an explicitly given stdin (i.e `!self.default_stdin`)

Comment (Joining7943) [reply src/uu/tail/src/args.rs:272]: This warnings should only be printed when you explicitely follow a redirected tty (among some special circumstances) to warn about the intention as being inefficient. Maybe also a look into gnu's implementation of tail regarding this warning explains this a little bit more.

File: src/uu/tail/src/follow/watch.rs (hunk 6 above)
Comment (Joining7943) [reply src/uu/tail/src/follow/watch.rs:483]: `settings.pid` only stores the initial configuration given by the user from the command line, but the observer acts upon the final value, for example after checking the pid support on the platform.

File: src/uu/tail/src/paths.rs (hunk 5 above)
Comment (Joining7943) [reply src/uu/tail/src/paths.rs:24]: :)

File: src/uu/tail/src/platform/unix.rs (hunk 4 above)
Comment (Joining7943) [reply src/uu/tail/src/platform/unix.rs:50]: Removing this function entirely then (together with platform::windows::stdin_is_pipe_or_fifo)? I haven't found a usage for it, but didn't want to remove it completely right away.

File: src/uu/tail/src/follow/watch.rs (hunk 6 above)
Comment (tertsdiepraam) [reply src/uu/tail/src/follow/watch.rs:483]: Ah gotcha, that makes sense

File: src/uu/tail/src/platform/unix.rs (hunk 4 above)
Comment (tertsdiepraam) [reply src/uu/tail/src/platform/unix.rs:50]: If it's not not used it can indeed be removed

File: src/uu/tail/src/follow/watch.rs (hunk 3 above)
Comment (tertsdiepraam) [reply src/uu/tail/src/follow/watch.rs:101]: Oops guess I start sounding like a broken record. I agree on this one, but take care not to introduce too many unnecessary changes.

File: src/uu/tail/src/args.rs (hunk 1 above)
Comment (Joining7943) [reply src/uu/tail/src/args.rs:295]: I like that. You're totally right, that way it looks like UResult, but using a UResult<Option<i32>> or alike is also a little bit ugly and obscure. Now that I look at VerificationResult it wasn't good choice of enum constants. Plain enums without value like VerificationResult::{Ok, CannotFollowByName, NoOutput} would have been better. Then the decision about what to do is moved into the calling function and it's easily extensible if needed. Would this solve your concerns?

File: src/uu/tail/src/args.rs (hunk 2 above)
Comment (tertsdiepraam) [reply src/uu/tail/src/args.rs:269]: >  but the order of the warnings would be mixed up then

In light of https://github.com/uutils/coreutils/issues/4000, I'm not super concerned about the warning order.

If you do want to keep the order, I would keep it simple and just change the `pid` in place.

File: src/uu/tail/src/args.rs (hunk 7 above)
Comment (tertsdiepraam) [reply src/uu/tail/src/args.rs:272]: > Maybe also a look into gnu's implementation of tail

I don't want to have to do that :) And I'm not speaking as a reviewer here, I'm fine with checking against GNU, but I think it should be obvious from the code why this is happening, so please add a similar explanation here.

File: src/uu/tail/src/follow/watch.rs (hunk 3 above)
Comment (Joining7943) [reply src/uu/tail/src/follow/watch.rs:101]: No problem :) I know what you mean, but this struct was introduced in a previous pr and I thought you agree on this change. I usually don't rename at all but this was a struct I've introduced myself...

File: src/uu/tail/src/platform/unix.rs (hunk 4 above)
Comment (Joining7943) [reply src/uu/tail/src/platform/unix.rs:50]: Ok

File: src/uu/tail/src/args.rs (hunk 1 above)
Comment (tertsdiepraam) [reply src/uu/tail/src/args.rs:295]: It'd be better I suppose, but since you don't have to do anything with the result except turn it into an error, I feel like just creating an error makes the most sense. In my opinion it's a situation where the additional type makes it more complicated instead of making it more clear. Either it's fine or we immediately exit with an error.

File: src/uu/tail/src/args.rs (hunk 2 above)
Comment (Joining7943) [reply src/uu/tail/src/args.rs:269]: #4000 is a cool issue. But keeping the order the way gnu's tail is producing them isn't a big task either. I'll try to keep the pid change in settings then.

File: src/uu/tail/src/args.rs (hunk 1 above)
Comment (Joining7943) [reply src/uu/tail/src/args.rs:290]: Sorry, haven't seen your last comment. I've pushed a version of VerificationResult I was talking about before.

File: src/uu/tail/src/args.rs (hunk 7 above)
Comment (Joining7943) [reply src/uu/tail/src/args.rs:273]: I haven't documented anything, right now. I'm going to do this.

File: src/uu/tail/src/args.rs (hunk 1 above)
Comment (Joining7943) [reply src/uu/tail/src/args.rs:290]: > Either it's fine or we immediately exit with an error.

Can't do this because we also exit without error. Maybe the updated version helps a little bit.

**Comment by Joining7943** [comment]
Sorry, that the tests for the warnings aren't integrated right now. But, I've ran the tests against the last 3 commits locally, so these changes are most probably working. We'll see this hopefully soon also in in the ci.

File: src/uu/tail/src/args.rs (hunk 1 above)
Comment (tertsdiepraam) [reply src/uu/tail/src/args.rs:290]: We can actually. `USimpleError`, `ExitCode` and friends all support exit code `0`. I admit it's a bit weird, but if you interpret `Err` as "exit the program with maybe an error" it makes sense.

> Sidenote: Once the `Try` trait stabilizes we might be able to create a custom `Result` or something that has a `exit_with_success` kinda variant, but it's not really a pressing issue.

Comment (Joining7943) [reply src/uu/tail/src/args.rs:290]: I don't really like that. We're exchanging something clear without hidden stuff against something obscure and weird. What's wrong with a simple and punctly enum thats coming with 0 costs? When I read Err(..) or USimpleError, I'm instantly thinking about an error and expect an error result but returning with 0 is definitely not an error.

Comment (tertsdiepraam) [reply src/uu/tail/src/args.rs:290]: Must be that I'm very familiar with all the `UResult` stuff cause I wrote it then. Feel free to leave it like it is then.

Comment (tertsdiepraam) [reply src/uu/tail/src/args.rs:290]: I got curious if we already did something like the error handling I suggested elsewhere. Turns out we do in exactly one spot, which actually should probably be a `show_warning!`:

https://github.com/uutils/coreutils/blob/0324a9480e7362d7436c0406331185d7b24adcad/src/uu/tr/src/tr.rs#L75-L82

So yeah, it's probably not how people tend to think about it.

> What's wrong with a simple and punctly enum thats coming with 0 costs?

There's one tiny cost, which is that the effect is decoupled from the check. Let's say I run the program and run into this error and I want to check where it came from. I then search the code for the error string and find that it's caused by some enum and now I have to find where that enum appears and I can find where my error originated. Again, not a big deal but this a tiny bit of additional abstraction. Arguably, however, one could also argue that we should go all in on the enum approach, not just for verification but all errors. Much like how other utils like `cat` or `ls` handle their errors.

Comment (Joining7943) [reply src/uu/tail/src/args.rs:290]: > Feel free to leave it like it is.

Ok.

> Arguably, we should go all in on the enum approach, not just for verification but all errors. 

I'm totally on your side. I like it a lot how rust made enums, types, matches etc. from functional style their own.

> ... but that's probably best left to another PR.

We could talk about it in #3952, if you like to (and when I'm ready with UChild and drawing some more commits out). I think we've then got a proper base for it there.

**Comment by Joining7943** [comment]
I've added all additional tests written during the refactoring of tail. These include the tests for the warnings. A lot of of these additional tests (besides the warnings tests) are currently (partly) disabled and only show the bugs and wrong behavior if enabled. I could fix many of them on the way of #3952 and they are therefore enabled there.

File: src/uu/tail/src/args.rs (hunk 7 above)
Comment (Joining7943) [reply src/uu/tail/src/args.rs:281]: done

**Comment by Joining7943** [comment]
@tertsdiepraam ping :) i think this pr would be ready for a final review.

File: src/uu/tail/src/args.rs (hunk 7 above)
Comment (tertsdiepraam) [reply src/uu/tail/src/args.rs:274]: So I looked into the GNU code and I think the `default_stdin` is wrong. The equivalent in their code is `found_hyphen` which is determined from the list of files. But if no files are given then this list is set to `"-"`, so `found_hyphen` will be true for when no `default_stdin` is given. I think `found_hyphen` is just exactly `self.has_stdin()`.

File: tests/by-util/test_tail.rs [hunk 8]
```diff
@@ -365,61 +379,6 @@ fn test_follow_stdin_name_retry() {
     }
 }
 
-#[test]
-#[cfg(target_os = "linux")]
-#[cfg(disable_until_fixed)]
-fn test_follow_stdin_explicit_indefinitely() {
```
Comment (tertsdiepraam) [inline tests/by-util/test_tail.rs:371]: I might have asked before, but I forgot. Why was this removed?

File: src/uu/tail/src/args.rs (hunk 7 above)
Comment (Joining7943) [reply src/uu/tail/src/args.rs]: You're right, `self.has_stdin()` is the exact equivalent of `found_hyphen`. I adjusted the if clause to match gnu's tail more closely.

File: tests/by-util/test_tail.rs (hunk 8 above)
Comment (Joining7943) [reply tests/by-util/test_tail.rs:371]: I've written more tests for this warning than this test would have tested if it were not disabled, so it can be removed in my opinion.

**Comment by Joining7943** [comment]
Are we ready, here?

File: src/uu/tail/src/paths.rs [hunk 9]
```diff
@@ -36,6 +26,7 @@ pub struct Input {
 }
 
 impl Input {
+    // TODO: from &str may be the better choice
```
Comment (tertsdiepraam) [inline src/uu/tail/src/paths.rs:29]: Not if the `display_name` field stays a `String`. Although `Into<String>` could work.

File: src/uu/tail/src/paths.rs [hunk 10]
```diff
@@ -89,6 +80,7 @@ impl Input {
     pub fn is_tailable(&self) -> bool {
         match &self.kind {
             InputKind::File(path) => path_is_tailable(path),
+            // TODO: shouldn't the default be true instead of false??
```
Comment (tertsdiepraam) [inline src/uu/tail/src/paths.rs:83]: Why? If we can't find the path, how can we tail it? Or do you want to print an error while tailing?

**Review by tertsdiepraam — COMMENTED** [review]
Code looks good, I just have some questions about the TODO's

File: src/uu/tail/src/paths.rs (hunk 10 above)
Comment (Joining7943) [reply src/uu/tail/src/paths.rs:83]: This method together with `Input::resolve` was removed from `Input` in the main refactoring pr, so this TODO can be ignored and I'm removing it. Sorry, I just quickly added this TODO at that time because I wanted to think about it, but later, and forgot to remove it from this pr.

File: src/uu/tail/src/paths.rs (hunk 9 above)
Comment (Joining7943) [reply src/uu/tail/src/paths.rs:29]: The field stays a `String`. This is no big change, I just thought accepting `&str` as parameter is the more general and popular approach because a `String` is coercing to `&str` when using a simple & on it. `&String` is also the type we get from `matches` in `Settings::from` when parsing the file arguments to `Input`s, and it seemed more natural to accept a reference as parameter in a `from` method.

**Review by tertsdiepraam — APPROVED** [review]
